dm/relay/binlog_writer.go

```diff
@@ -61,6 +74,71 @@ func NewBinlogWriter(logger log.Logger, relayDir string) *BinlogWriter {
 	return &BinlogWriter{
 		logger:   logger,
 		relayDir: relayDir,
+		errCh:    make(chan error, 1),
+	}
+}
+
+// run starts the binlog writer.
+func (w *BinlogWriter) run() {
+	var (
+		buf       = &bytes.Buffer{}
+		errOccurs bool
+	)
+
+	// writeToFile writes buffer to file
+	writeToFile := func() {
+		if buf.Len() == 0 {
+			return
+		}
+
+		w.mu.Lock()
+		defer w.mu.Unlock()
+		if w.file == nil {
+			select {
+			case w.errCh <- terror.ErrRelayWriterNotOpened.Delegate(errors.New("file not opened")):
+			default:
+			}
+			errOccurs = true
+			return
+		}
+		n, err := w.file.Write(buf.Bytes())
+		if err != nil {
+			select {
+			case w.errCh <- terror.ErrBinlogWriterWriteDataLen.Delegate(err, n):
+			default:
+			}
+			errOccurs = true
+			return
+		}
+		buf.Reset()
+	}
+
+	for {
+		select {
+		case bs, ok := <-w.input:
```

Comment:
when we can read something from the buffered channel, we can read all of them then do a flush. This can achieve low latency because no `waitTime` or `bufferSize` is needed

Reply:
I have try with in [dc9edc5](https://github.com/pingcap/tiflow/pull/7580/commits/dc9edc540e47abbfea1407b5e55b5517364af2c4) originally, but the performance is even worse than the original....
```
for bs := <- range w.input{
     ...
     if len(w.input)==0{
          flush
     }
}
```

Reply:
😂 ptal @D3Hunter

Reply:
Does @lance6716 mean reading and writing as long as the channel is full?

Reply:
I think we should be able to do a block reading on channel, and when the goroutine continues it reads all data from the channel to flush them. This is more adaptive.

Maybe we need to take a look at profling to see why this idea does not work.

Reply:
performance issue fixed after remove the lock in relay writer [20c870a](https://github.com/pingcap/tiflow/pull/7580/commits/20c870a54ddc43b52dbc5567c06fb7b8eec3cbff)